cypress-io/cypress
# feat: Debug run navigation

- Closes https://github.com/cypress-io/cypress/issues/25899
- Closes https://github.com/cypress-io/cypress/issues/26018

### Additional details


There are two main parts to this PR:
1. addition of the new `DebugRunNavigation` component and its child `DebugRunNavigationRow` component
2. Updates to the `RelevantRunsDataSource` and `RelevantRunSpecsDataSource` to support it

The other changes all involve refactoring various parts of the code for reuse, and some general UI fixes/improvements that were found will testing this feature.

### Steps to test

You can see various states of this component by opening the App for existing projects (like our own Cypress project), but the easiest way to test will be to have your own Cloud project that you can easily record local runs against.

Scenarios to test:
- Start with a project with no runs.  When connected to the cloud you should see the existing empty state to record a run
- Record the first run in the project, the Debug page should update to show the running build and no run navigation should be present since it is the only run
- Make a commit in your project and then record a new run
- The debug page should update to show the run navigation component that indicates that the new build is running
![image](https://user-images.githubusercontent.com/1702361/227805329-5603e5c5-810e-4508-bb2f-322e9482ff2c.png)
  - The component will **not** show the expandable component since there is only one run to switch to
  - Click the "Switch to the newest run" button to view the running build
  - The component will update to show you are on the most recent build and you should be able to expand the component and switch back to the previous running build
<img width="1011" alt="image" src="https://user-images.githubusercontent.com/1702361/227805689-f779120a-952c-4ba9-870b-1b1898cb9b25.png">

- When the running build completes, the component will be removed since there is only one build for the most recent commit with a completed build and no newer builds
- Record another run **without** making a new commit
  - The run navigation component will appear again and show the running build
  - You can switch back and forth between the runs 
- When the running build completes, the navigation component will remain.  This is because there are multiple builds for the same most relevant commit with builds.  

### How has the user experience changed?


The Debug page will now show a new component at the top when scenarios allow for changing between different runs for debugging.

See the linked issue https://github.com/cypress-io/cypress/issues/25899 for screenshots and details.

### PR Tasks


- [] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
